move/squid/sources/squid/versioned/squid_v0.move

```diff
@@ -0,0 +1,161 @@
+        coin_bag: coin_bag::new(ctx),
+        version_control,
+    }
+}
+
+public(package) fun channel(self: &SquidV0): &Channel {
+    &self.channel
+}
+
+public(package) fun version_control(self: &SquidV0): &VersionControl {
+    &self.version_control
+}
+
+public(package) fun coin_bag_mut(self: &mut SquidV0): &mut CoinBag {
+    &mut self.coin_bag
+}
+
+public fun start_swap<T>(
+    self: &mut SquidV0,
+    its: &mut ITS,
+    approved_message: ApprovedMessage,
+    clock: &Clock,
+    ctx: &mut TxContext,
+): SwapInfo {
+    let (_, _, data, coin) = service::receive_interchain_transfer_with_data<T>(
+        its,
+        approved_message,
+        self.channel(),
+        clock,
+        ctx,
+    );
+    let mut swap_info = swap_info::new(data, ctx);
+    swap_info.coin_bag().store_estimate<T>(coin.value());
+    swap_info.coin_bag().store_balance(coin.into_balance());
+    swap_info
+}
+
+#[test_only]
+public fun new_for_testing(ctx: &mut TxContext): SquidV0 {
+    SquidV0 {
+        channel: channel::new(ctx),
+        coin_bag: coin_bag::new(ctx),
+        version_control: version_control::new(vector[]),
+    }
+}
+
+#[test_only]
+use its::coin::COIN;
+
+#[test]
+fun test_start_swap() {
+    let ctx = &mut tx_context::dummy();
+    let clock = sui::clock::create_for_testing(ctx);
+    let mut its = its::its::new_for_testing();
+    let mut squid = new_for_testing(ctx);
+
+    let coin_info = its::coin_info::from_info<COIN>(
+        std::string::utf8(b"Name"),
+        std::ascii::string(b"Symbol"),
+        10,
+        12,
+    );
+
+    let amount = 1234;
+    let data = std::bcs::to_bytes(&vector<vector<u8>>[]);
+    let coin_management = its::coin_management::new_locked();
+    let coin = sui::coin::mint_for_testing<COIN>(amount, ctx);
+
+    let token_id = its::service::register_coin(
+        &mut its,
+        coin_info,
+        coin_management,
+    );
+
+    // This gives some coin to the service.
+    let interchain_transfer_ticket = service::prepare_interchain_transfer(
+        token_id,
+        coin,
+        std::ascii::string(b"Chain Name"),
+        b"Destination Address",
+        b"",
+        &squid.channel,
+    );
+    sui::test_utils::destroy(
+        service::send_interchain_transfer(
+            &mut its,
+            interchain_transfer_ticket,
+            &clock,
+        ),
+    );
+
+    let source_chain = std::ascii::string(b"Chain Name");
+    let message_id = std::ascii::string(b"Message Id");
+    let message_source_address = std::ascii::string(b"Address");
+    let its_source_address = b"Source Address";
+
+    let destination_address = squid.channel().to_address();
+
+    let mut writer = abi::abi::new_writer(6);
+    writer
+        .write_u256(0)
+        .write_u256(token_id.to_u256())
+        .write_bytes(its_source_address)
+        .write_bytes(destination_address.to_bytes())
+        .write_u256((amount as u256))
+        .write_bytes(data);
+    let payload = writer.into_bytes();
```

Comment:
```suggestion
    let payload = abi::abi::new_writer(6)
        .write_u256(0)
        .write_u256(token_id.to_u256())
        .write_bytes(its_source_address)
        .write_bytes(destination_address.to_bytes())
        .write_u256((amount as u256))
        .write_bytes(data)
        .into_bytes();
```
since writer isn't being used anywhere else

Reply:
This does not work:
- `writer.write_*` returns `&mut AbiWriter`
- `writer.into_bytes()` takes `AbiWriter`
so you cannot use one for the other.